### DISCUSSION TITLE
Duvida sobre FluenteMap com Primitive Obsession

### DISCUSSION THREAD

**Discussion by wezmoreira** [balta-io/forum#296] [Tira Dúvidas] [answered]
Estava assistindo a aula sobre Primitive Obsession e tentei replicar em uma Api esse conceito, porém ao tentar executar as migrações ele da um erro:

> The expression 'x => x.Email.Address' is not a valid member access expression. The expression should represent a simple property or field access: 't => t.MyProperty'. (Parameter 'memberAccessExpression')

Eu não sei se foi eu que fiz alguma coisa errada, ou se é um jeito diferente pra realizar o mapeamento. Estou usando o mapeamento que foi apresentada no curso "2811".

**Comment by andrebaltieri** [comment, marked as answer]
Boa tarde, @wezmoreira, como vai? Qual trecho do código apresentou este erro?

**Reply by wezmoreira** [reply]
Tenho minha entidade User que tem uma propriedade Email que é um VO e no email tem a propriedade Address, por exemplo:

```
public class User : Entity
{
  public EmailVo Email { get; private set; }
}
```

```
public class EmailVo 
{
    public string Address { get; private set; }
}
```
Então tento gerar as migrações para o banco com o seguinte mapeamento da entidade User:

```
public class UserMap : IEntityTypeConfiguration<User>
{
    public void Configure(EntityTypeBuilder<User> builder)
    {
         builder.Property(x => x.Email.Address)
            .IsRequired()
            .HasColumnName("Email_Address")
            .HasColumnType("NVARCHAR")
            .HasMaxLength(120);
    }
}
```
Nesse momento surge o erro que eu citei

> The expression 'x => x.Email.Address' is not a valid member access expression. The expression should represent a simple property or field access: 't => t.MyProperty'. (Parameter 'memberAccessExpression')

E minha dúvida é se eu estou fazendo o mapeamento do jeito errado ou se tem mais alguma configuração que preciso fazer, pois pelo que eu pesquisei ele não consegue realizar o mapeamento acessando uma propriedade dentro de outra propriedade.

**Reply by andrebaltieri** [reply]
Bom dia, @wezmoreira este erro ocorre por que a propriedade `Email.Address` é complexa e não pode ser mapeada assim.
É preciso utilizar o `OwnsOne`:
```csharp
builder.OwnsOne(x => x.Email)
    .Property(x => x.Address)
    .HasColumnName("Email")
    .IsRequired(true);
``` 

Aqui tem mais detalhes: https://balta.io/blog/ef-core-value-objects
